include/RevTracer.h

```diff
@@ -0,0 +1,229 @@
+//
+// _RevTracer_h_
+//
+// Copyright (C) 2017-2023 Tactical Computing Laboratories, LLC
+// All Rights Reserved
+// contact@tactcomplabs.com
+//
+// See LICENSE in the top level directory for licensing details
+//
+//
+
+#ifndef _SST_REVCPU_REVTRACER_H_
+#define _SST_REVCPU_REVTRACER_H_
+
+// -- SST Headers
+#include "SST.h"
+
+// -- Standard Headers
+#include <cstdint>
+#include <ostream>
+#include <string>
+#include <iostream>
+#include <map>
+#include <vector>
+
+// -- Rev Headers
+
+// Integrated Disassembler (toolchain dependent)
+#ifdef REV_USE_SPIKE
+// sst and spike cause redefinition warning for macro UNUSED.
+// sst: #define UNUSED(x) x __attribute__((unused))
+// spike: # define UNUSED __attribute__ ((unused))
+// TODO can these be uniquified?
+#ifdef __GNUC__
+#undef UNUSED
+#endif
+#include "riscv/disasm.h"
+#endif
+
+// Tracing macros
+#ifndef NO_REV_TRACER
+#define TRACE_REG_READ(R,V)  { if (Tracer) Tracer->regRead(  (uint8_t) (R),(uint64_t) (V) ); }
+#define TRACE_REG_WRITE(R,V) { if (Tracer) Tracer->regWrite( (uint8_t) (R),(uint64_t) (V) ); }
+#define TRACE_PC_WRITE(PC)   { if (Tracer) Tracer->pcWrite( (uint64_t) (PC) ); }
+#define TRACE_MEM_WRITE(ADR, LEN, DATA) { if (Tracer) Tracer->memWrite( (ADR), (LEN), (DATA) ); }
+#define TRACE_MEM_READ(ADR, LEN, DATA)  { if (Tracer) Tracer->memRead(  (ADR), (LEN), (DATA) ); }
+#else
+#define TRACE_REG_READ(R,V)
+#define TRACE_REG_WRITE(R,V)
+#define TRACE_PC_WRITE(PC)
+#define TRACE_MEM_WRITE(ADR, LEN, DATA)
+#define TRACE_MEM_READ(ADR, LEN, DATA)
+#endif
+
+namespace SST{
+  namespace RevCPU{
+
+  // Tracing controls are using custom hint SLTI rd = x0
+  // See unpriv-isa-asiidoc.pdf Section 2.9 HINT Instruction
+  // OP     RD    Space  Assembly           Encoding 
+  // SLTI   rd=x0 2^17   slti  x0, x0, ?    0x00?02013  //
+  // SLTIU  rd=x0 2^17   sltiu x0, x0, ?    0x00?03013  //
+  // SLLI   rd=x0 2^10   slli  x0, x0, ?    0x00?01013  // Default
+  // SRLI   rd=x0 2^10   srli  x0, x0, ?    0x00?05013  //
+  // SRAI   rd=x0 2^10   srai  x0, x0, ?    0x40?05013  //
+  // SLT    rd=x0 2^10   slt   x0, x0, x?   0x00?02033  // Not supported
+  // SLTU   rd=x0 2^10   sltu  x0, x0, x?   0x00?03033  // Not supported
+  const std::map<std::string, uint32_t> s2op {
+    {"slti",  0x00002013}, 
+    {"sltiu", 0x00003013},
+    {"slli",  0x00001013},
+    {"srli",  0x00005013},
+    {"srai",  0x40005013}
```

Comment:
A trailing comma on the last entry is allowed and encouraged, to ease adding new entries or shuffling existing entries, or for automatic code generation.